rust/vapid/src/lib.rs

```diff
@@ -24,116 +26,107 @@
 //!
 //! // The result will contain the `Authorization:` header. How you inject this into your
 //! // request is left as an exercise.
-//! let authorization_header = sign(key, &mut claims)?;
+//! let authorization_header = sign(my_key, &mut claims).unwrap();
 //!
 //! ```
 
 extern crate base64;
+#[macro_use()]
 extern crate openssl;
 extern crate serde_json;
 extern crate time;
 
 use std::collections::HashMap;
+use std::hash::BuildHasher;
 use std::fs::File;
 use std::io::Read;
 use std::io::Write;
 use std::path::Path;
 
 use openssl::bn::BigNumContext;
+use openssl::ec::{self, EcKey};
 use openssl::hash::MessageDigest;
-use openssl::ec;
 use openssl::nid;
+use openssl::pkey::{PKey, Private, Public};
 use openssl::sign::{Signer, Verifier};
-use openssl::pkey;
 
-pub struct Key(ec::EcKey);
+mod error;
+pub struct Key {
+    key: EcKey<Private>,
+}
 
 /// a Key is a helper for creating or using a VAPID EC key.
 ///
 /// Vapid Keys are always Prime256v1 EC keys.
 ///
 impl Key {
     fn name() -> nid::Nid {
-        nid::X9_62_PRIME256V1
-    }
-
-    fn group() -> ec::EcGroup {
-        ec::EcGroup::from_curve_name(Key::name()).expect("EC Prime256v1 curve not supported")
+        nid::Nid::X9_62_PRIME256V1
     }
 
     /// Read a VAPID private key stored in `path`
-    pub fn from_pem<P>(path: P) -> Result<Key, String>
-        where P: AsRef<Path>{
+    pub fn from_pem<P>(path: P) -> error::VapidResult<Key>
+    where
+        P: AsRef<Path>,
+    {
         let mut pem_data = Vec::new();
```

Comment:
They added a few nice helpers for reading files last year, this one'll kill a couple lines:

```suggestion
         let mut pem_data = fs::read(&path)?;
```